command.go

```diff
@@ -130,101 +143,148 @@ func (srv *Server) handleMessageSizeExceeded(reader *buffer.Reader, writer *buff
+			// error, or roll back if error). Then a ReadyForQuery response is
+			// issued. The purpose of Sync is to provide a resynchronization point
+			// for error recovery. When an error is detected while processing any
+			// extended-query message, the backend issues ErrorResponse, then reads
+			// and discards messages until a Sync is reached, then issues
+			// ReadyForQuery and returns to normal message processing. (But note
+			// that no skipping occurs if an error is detected while processing Sync
+			// — this ensures that there is one and only one ReadyForQuery sent for
+			// each Sync.)
+			// https://www.postgresql.org/docs/current/protocol-flow.html#PROTOCOL-FLOW-EXT-QUERY
+			return readyForQuery(writer, types.ServerIdle)
+		case types.ClientBind:
+			return srv.handleBind(ctx, reader, writer)
+		case types.ClientFlush:
+			// TODO: Flush all remaining rows inside connection buffer if
+			// any are remaining.
+			//
+			// The Flush message does not cause any specific
+			// output to be generated, but forces the backend to deliver any data
+			// pending in its output buffers. A Flush must be sent after any
+			// extended-query command except Sync, if the frontend wishes to examine
+			// the results of that command before issuing more commands. Without
+			// Flush, messages returned by the backend will be combined into the
+			// minimum possible number of packets to minimize network overhead.
+			// https://www.postgresql.org/docs/current/protocol-flow.html#PROTOCOL-FLOW-EXT-QUERY
+			return nil
+		case types.ClientCopyData, types.ClientCopyDone, types.ClientCopyFail:
+			// We're supposed to ignore these messages, per the protocol spec. This
+			// state will happen when an error occurs on the server-side during a copy
+			// operation: the server will send an error and a ready message back to
+			// the client, and must then ignore further copy messages. See:
+			// https://github.com/postgres/postgres/blob/6e1dd2773eb60a6ab87b27b8d9391b756e904ac3/src/backend/tcop/postgres.c#L4295
+			return nil
+		case types.ClientClose:
+			// TODO: close the statement or portal
+			writer.Start(types.ServerCloseComplete) //nolint:errcheck
+			writer.End()                            //nolint:errcheck
+			return nil
+		case types.ClientTerminate:
+			err := srv.handleConnTerminate(ctx)
+			if err != nil {
+				return err
+			}
+
+			err = conn.Close()
+			if err != nil {
+				return err
+			}
+
+			return io.EOF
+		default:
+			return ErrorCode(writer, NewErrUnimplementedMessageType(t))
 		}
+	}
+}
 
-		err = conn.Close()
-		if err != nil {
-			return err
+// copyDataFn returns the default [CopyDataFn] function.
+func (srv *Server) copyDataFn(reader *buffer.Reader, writer *buffer.Writer) CopyDataFn {
+	return func(ctx context.Context) ([]byte, error) {
+		var results []byte
+		err := srv.commandLoop(ctx, reader, writer, srv.handleCopyInCommand(func(r []byte) { results = r }))
```

Comment:
I had the same thought to turn it into a reader. So good call. Done.

Reply:
I agree this is a bit cumbersome. I'm not sure that an io.Writer or goroutine would help much, though.

We're not actually reading anything extra into memory as it is.  The `results`  value just gets the contents of `reader.Msg`, which is already completely in memory.

The reason for this awkward callback is to facilitate a common function signature for `handleCommand` and `handleCopyInCommand`, so that they can both be passed to `commandLoop` (formerly part of `consumeCommands`.

Reply:
I've refactored this bit now, to be quite a bit simpler, I think. I'm curious to hear your thoughts on how it looks now.

Reply:
Interesting approach, it is indeed a bit simpler. Got a few ideas I was thinking about which I would play with tonight to check if we could avoid having to copy the buffer.

Reply:
What if we pass the content length to the handler function. This way we could stream the incoming bytes to the returned io.Reader and terminate the io.Reader when reaching the total byte length. It might make sense to take over the buffered reader in case of a copy command in order to avoid the `ErrMessageSizeExceeded` error being thrown for larger payloads.

Reply:
Which content length are you referring to?  For the individual CopyData packets?  Or the overall length of the copy data?  I don't think the latter is communicated to the server (or even known by the client in many cases).  But maybe I'm overlooking something obvious.

Reply:
I am not able to test this at the moment, typed messages within the psql wire protocol have a specific length. We need to check whether the message length is send for the entire copydata message or whether the messages are split in chunks. The message length is fetched over here: https://github.com/jeroenrinzema/psql-wire/blob/main/pkg/buffer/reader.go#L125-L131

Reply:
I still don't quite follow.

A single COPY IN is typicall comprised of multile CopyData packets.

I don't think there's any way to know the overall length of the entire stream of data to be copied in, until the CopyDone packet is received.

Reply:
Apologies for the late response I have been sick over the past week. Good point, I missed that within the PSQL wire protocol implementation. The way how we could do it is as following:

```go
type copyDataReader struct {
	reader *buffer.Reader
	size   int
}

func (r *copyDataReader) Read(p []byte) (int, error) {
	if r.size == 0 {
		_, err := r.reader.ReadType()
		if err != nil {
			return -1, err
		}

		r.size, err = r.reader.ReadMsgSize()
		if err != nil {
			return -1, err
		}
	}

	n, err := io.ReadFull(r.reader.Buffer, p)
	r.size -= n
	return n, err
}
```

We need to expose a bit more internal logic within the buffered reader but this would allow us to bypass copying the buffer twice. The reader directly interacts with the wire protocol and passes the given buffer by the reader to `io.ReadFull`.  I have just opened a PR exposing the methods used in the example above: https://github.com/jeroenrinzema/psql-wire/pull/74.

Reply:
@flimzy I have implemented the changes discussed above and have included a simple binary column reader for the copy reader. Feel free to check it out!